packages/remix-dev/compiler/manifest.ts

```diff
@@ -106,13 +106,23 @@ export async function create({
     }
   }
 
+  // At this point, `routes` is sorted by the esbuild metafile order, but we
+  // want them exposed in the order they came in from the config
+  let order: Record<string, number> = Object.keys(config.routes).reduce(
+    (acc, routeId, idx) => Object.assign(acc, { [routeId]: idx }),
+    {}
+  );
+  let sortedRoutes = Object.keys(routes)
+    .sort((a, b) => (order[a] < order[b] ? -1 : order[a] > order[b] ? 1 : 0))
+    .reduce((acc, id) => Object.assign(acc, { [id]: routes[id] }), {});
+
```

Comment:
Vite does not have this issue because it generates the manifest in the order it gets from `remixConfig.routes` so the order is already correct there after the fix to `flat-routes.ts` below.